pkg/timeutil/window.go

```diff
@@ -0,0 +1,181 @@
+package timeutil
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+)
+
+var (
+	nilDay = time.Weekday(-1)
+
+	shortDayNames = map[string]time.Weekday{
+		"sun": time.Sunday,
+		"mon": time.Monday,
+		"tue": time.Tuesday,
+		"wed": time.Wednesday,
+		"thu": time.Thursday,
+		"fri": time.Friday,
+		"sat": time.Saturday,
+	}
+)
+
+func d2s(d time.Weekday) string {
+	s := d.String()
+	if len(s) >= 3 {
+		return s[:3]
+	}
+	return s
+}
+
+type dayTime struct {
+	day     time.Weekday // day of week, -1 if unset
+	minutes int          // minutes into day
+}
+
+func parseTime(s string) (hh, mm int, err error) {
+	spl := strings.Split(s, ":")
+	if len(spl) != 2 {
+		err = errors.New("malformed HH:MM specification")
+		return
+	}
+
+	hh, err = strconv.Atoi(spl[0])
+	if err != nil {
+		err = fmt.Errorf("malformed hour %q: %s", spl[0], err)
+		return
+	}
+
+	mm, err = strconv.Atoi(spl[1])
+	if err != nil {
+		err = fmt.Errorf("malformed minute %q: %s", spl[1], err)
+		return
+	}
+
+	return
+}
+
+func parseDayTime(s string) (dayTime, error) {
+	dt := dayTime{nilDay, 0}
+	spl := strings.Fields(s)
+
+	switch len(spl) {
+	default:
+		fallthrough
+	case 0:
+		return dt, errors.New("empty time specification")
+	case 2:
+		// 2 fields, day hh:mm
+		day, ok := shortDayNames[strings.ToLower(spl[0])]
+		if !ok {
+			return dt, fmt.Errorf("invalid weekday %q", spl[0])
+		}
+
+		dt.day = day
+
+		spl = spl[1:]
+		fallthrough
+	case 1:
+		// 1 field, must be time.
+		hh, mm, err := parseTime(spl[0])
+		if err != nil {
+			return dt, err
+		}
+
+		dt.minutes = (hh * 60) + mm
+		return dt, nil
+	}
+}
+
+type Window struct {
+	start dayTime
+	end   dayTime
+}
+
+func ParseWindow(start string, end string) (*Window, error) {
+	s, err := parseDayTime(start)
+	if err != nil {
+		return nil, fmt.Errorf("start time %q: %s", start, err)
+	}
+
+	e, err := parseDayTime(end)
+	if err != nil {
+		return nil, fmt.Errorf("end time %q: %s", end, err)
+	}
+
+	// check weekday is present in both or neither
+	if (s.day == nilDay) != (e.day == nilDay) {
+		return nil, errors.New("day of week must be present in both or neither time spec")
+	}
+
+	return &Window{s, e}, nil
+}
+
+func (w Window) String() string {
+	var s string
+
+	for i, d := range []dayTime{w.start, w.end} {
+		if d.day != nilDay {
+			s += d2s(d.day) + " "
+		}
+
+		hh, mm := d.minutes/60, d.minutes%60
+
+		s += fmt.Sprintf("%02d:%02d", hh, mm)
+
+		if i == 0 {
+			s += " - "
+		}
+	}
+
+	return s
+}
+
+// In returns true if t is within the window, or false if not.
+func (w *Window) In(t time.Time) bool {
+	wd := t.Weekday()
+	hh, mm, _ := t.Clock()
+	minutes := (hh * 60) + mm
+
+	switch {
+	case w.start.day == nilDay || (w.start.day == wd && w.start.day == w.end.day):
```

Comment:
this case is actually bugged. if you specify wed 00:00 - wed 01:00, then it will return true on any day the time falls between 00:00 and 01:00, not just wed.